src/main/java/io/cdap/plugin/servicenow/sink/model/APIResponse.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.servicenow.sink.model;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Model class for API Response from ServiceNow Table APIs
+ */
+public class APIResponse {
+  
+  private List<Map<String, String>> result;
```

Comment:
How many results will the response have, and how many fields can each result have?

I am concerned about reading the entire response into memory instead of reading it in a streaming fashion.

Reply:
this should be final (see below comment about avoiding setter methods)

Reply:
> How many results will the response have, and how many fields can each result have?
> 
> I am concerned about reading the entire response into memory instead of reading it in a streaming fashion.

The ServiceNow Table API response contains a single `result` array. The result array can contain up to 5000 [PAGE_SIZE] records. Each record contains the fields (column names) of the requested table along with their values. The number of fields vary from table to table. Can be as low as ~10 or as high as ~50.

Sample Response Below for sys_user table.
```
{
    "result": [
        {
            "calendar_integration": "1",
            "country": "",
            "last_login_time": "2019-04-05 22:16:30",
            "hashed_user_id": "fdcbd2f17b127236728959a60d57a50338ba8d115adbf14c3d7dffd7dc3ac1fe",
            "source": "",
            "sys_updated_on": "2022-10-04 06:53:43",
            "building": "",
            "web_service_access_only": "false",
            "notification": "2",
            "enable_multifactor_authn": "false",
            "sys_updated_by": "system",
            "sys_created_on": "2019-04-05 21:09:12",
            "sys_domain": {
                "link": "https://ven05127.service-now.com/api/now/table/sys_user_group/global",
                "value": "global"
            },
            "state": "",
            "vip": "false",
            "sys_created_by": "admin",
            "zip": "",
            "home_phone": "",
            "time_format": "",
            "last_login": "2019-04-05",
            "active": "true",
            "sys_domain_path": "/",
            "cost_center": "",
            "phone": "",
            "name": "survey user",
            "employee_number": "",
            "gender": "",
            "city": "",
            "failed_attempts": "0",
            "user_name": "survey.user",
            "title": "",
            "sys_class_name": "sys_user",
            "sys_id": "005d500b536073005e0addeeff7b12f4",
            "internal_integration_user": "false",
            "ldap_server": "",
            "mobile_phone": "",
            "street": "",
            "company": "",
            "department": "",
            "first_name": "survey",
            "email": "survey.user@email.com",
            "introduction": "",
            "preferred_language": "",
            "manager": "",
            "sys_mod_count": "2",
            "last_name": "user",
            "photo": "",
            "avatar": "",
            "middle_name": "",
            "sys_tags": "",
            "time_zone": "",
            "schedule": "",
            "date_format": "",
            "location": ""
        },
        {
            "calendar_integration": "1",
            "country": "US",
            "last_login_time": "",
            "hashed_user_id": "30501c5b4b720ad4fd33232a98f83c21423222e5ce83594e36bf5c657b845431",
            "source": "null",
            "sys_updated_on": "2022-10-06 12:51:33",
            "building": "",
            "web_service_access_only": "false",
            "notification": "2",
            "enable_multifactor_authn": "false",
            "sys_updated_by": "pipeline.user.1",
            "sys_created_on": "2022-10-06 12:51:33",
            "sys_domain": {
                "link": "https://ven05127.service-now.com/api/now/table/sys_user_group/global",
                "value": "global"
            },
            "state": "Massachusetts",
            "vip": "false",
            "sys_created_by": "pipeline.user.1",
            "zip": "2116",
            "home_phone": "",
            "time_format": "",
            "last_login": "",
            "active": "true",
            "sys_domain_path": "/",
            "cost_center": "",
            "phone": "(732) 634-7777",
            "name": "Farmers Insurance ",
            "employee_number": "",
            "gender": "",
            "city": "Boston",
            "failed_attempts": "",
            "user_name": "",
            "title": "",
            "sys_class_name": "sys_user",
            "sys_id": "009e96f287d29110a91edb583cbb3586",
            "internal_integration_user": "false",
            "ldap_server": "",
            "mobile_phone": "",
            "street": "75 Park Plz",
            "company": "",
            "department": "",
            "first_name": "Farmers Insurance",
            "email": "null",
            "introduction": "",
            "preferred_language": "",
            "manager": "",
            "sys_mod_count": "0",
            "last_name": "",
            "photo": "",
            "avatar": "",
            "middle_name": "",
            "sys_tags": "",
            "time_zone": "",
            "schedule": "",
            "date_format": "",
            "location": ""
        }
    ]
}
```


> this should be final (see below comment about avoiding setter methods)
 `result` is now declared as a `final` variable. cbfb212c77242fe61b2ef14c690a9079b6a9e22f

Reply:
I don't see page_size being sent anywhere, is that a request property or something else? Also, does that mean the response is actually paginated? I don't see the code going through pages either, it just makes a single request. If the response is paginated, this plugin is definitely not ready for GA, as it will not be able to read anything after the first page.

Reply:
The PAGE_SIZE is passed here. https://github.com/cloudsufi/servicenow-plugins/blob/ad60671e737cd75675c454c195419e50d7160720/src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java#L102

```
restApi.fetchTableRecordsRetryableMode(tableName, pluginConf.getValueType(), pluginConf.getStartDate(),
                                                     pluginConf.getEndDate(), split.getOffset(),
                                                     ServiceNowConstants.PAGE_SIZE);
```

It later becomes part of the request object 

```
public ServiceNowTableAPIRequestBuilder setLimit(int limit) {
    this.parameters.put("sysparm_limit", String.valueOf(limit));
    return this;
}
```

Reply:
discussed offline. To close the loop here, the record readers are not paging through the results. Instead, the inputformat creates splits based off the total number of table rows, then sets an offset for each split. So each individual record reader reads one page by design.